### ISSUE TITLE
Slow local usb backup / high io load

### ISSUE THREAD

**Issue by otto802** [restic/restic#1555]
## Output of `restic version`
    restic 0.8.1
    compiled with go1.9.2 on linux/amd64

## How did you run restic exactly?

    export RESTIC_PASSWORD="123"
    export RESTIC_REPOSITORY="/mnt/usb/backup/restic"
    
    /opt/restic/restic backup --files-from /opt/restic/usb/restic_backup_dirs --exclude-file         /opt/restic/usb/restic_backup_exclude
 

## What backend/server/service did you use to store the repository?

* local mounted usb 3.0 drive (4TB) formatted with ext4
* Ubuntu 16.04.3 LTS

## Expected behavior

* higher throughput, lower system load

## Actual behavior

- backup of ~2TB mainly big video files 200GB smaller files (documents etc)
- current output of restic:

    password is correct
    scan [/data/data0]
    scanned 53385 directories, 828226 files in 0:22
    [40:34:48] 67.83%  10.637 MiB/s  1.482 TiB / 2.185 TiB  878217 / 881611 items  0 errors  ETA 19:14:47

- current load average:
 11:55:00 up 5 days,  1:59,  1 user,  load average: 33.42, 33.27, 33.08
- top output
%Cpu(s):  2.8 us,  0.3 sy,  0.0 ni, 24.2 id, 72.8 wa,  0.0 hi,  0.0 si,  0.0 st
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 7407 root      20   0 2015492 1.717g   8452 S 128.9 11.2   1638:24 restic

- the system is responsive
- but the backup is really slow here, expected it 3-5 times faster (aka the USB drive speed) ;-)
- I will test the USB drive speed afterwards I won't cancel the currently running backup for now

## Do you have any idea what may have caused this?

not really but something seem very io/cpu intensive. Maybe related to this bug: https://github.com/restic/restic/issues/1145

## Did restic help you or made you happy in any way?

yes :-), I use it successfully to backup (different files as here) to google cloud storage (with expected speed).

**Comment by otto802** [comment]
my backup just finished:

```
root@enterprise:/opt/restic/usb# ./restic_backup.sh
password is correct
scan [/data/data0]
scanned 53385 directories, 828226 files in 0:22
[63:14:24] 100.00%  10.078 MiB/s  2.188 TiB / 2.185 TiB  881632 / 881611 items  0 errors  ETA 0:00
duration: 63:14:24, 10.06MiB/s
snapshot 0f44880c saved
```

**Comment by fd0** [comment]
Hey, thanks for the report. The bandwidth that is displayed by restic is not the writing speed to the backend, but rather the reading/processing speed. At the moment, the archiver code does too many things in parallel (hence the high load, which is not necessarily bad), we're going to address that with the new archiver code (there's an open PR for it).

I find that the resulting processed bandwidth of ~10MiB/s is suspiciously close to what I usually achieved with USB 2 devices. Can you please check if the device maybe is in USB 2 mode?

Apart from that, did you check which IO load occurred? Is the reading side saturated, or the writing side?

**Comment by fd0** [comment]
Hm, which device is that exactly? The new >3TB hard discs may already use [SMR](https://en.wikipedia.org/wiki/Shingled_magnetic_recording), that may be a factor.

**Comment by fd0** [comment]
In order to test the speed, use a subdir of the data that you saved with restic, which is much larger than your RAM, then do something along the lines of:
```
$ tar c /data/data0 | pv > /mnt/usb/backup/testfile.tar
```

**Comment by otto802** [comment]
thanks for your reply. Its a new Seagate M3. Its plugged into a USB3 port but actually I don't know if its running in USB3 mode.

```
[538086.827385] usb 4-2: new SuperSpeed USB device number 2 using xhci_hcd
[538086.844252] usb 4-2: New USB device found, idVendor=0bc2, idProduct=61b7
[538086.844257] usb 4-2: New USB device strings: Mfr=2, Product=3, SerialNumber=1
[538086.844260] usb 4-2: Product: M3 Portable
[538086.844262] usb 4-2: Manufacturer: Seagate
[538086.844264] usb 4-2: SerialNumber: NM1452AJ
```

Just checked dmesg... during the backup there where a lot of strange messages:

```
[517944.248244] sd 7:0:0:0: [sdd] tag#20 CDB: Write(16) 8a 00 00 00 00 00 e8 c5 bd d0 00 00 00 08 00 00
[517944.248253] scsi host7: uas_eh_bus_reset_handler start
[517944.255506] sd 7:0:0:0: [sdd] tag#0 uas_zap_pending 0 uas-tag 1 inflight: CMD
[517944.255512] sd 7:0:0:0: [sdd] tag#0 CDB: Write(16) 8a 00 00 00 00 00 34 03 61 00 00 00 04 00 00 00
[517944.255516] sd 7:0:0:0: [sdd] tag#1 uas_zap_pending 0 uas-tag 2 inflight: CMD
[517944.255519] sd 7:0:0:0: [sdd] tag#1 CDB: Write(16) 8a 00 00 00 00 00 34 03 65 00 00 00 04 00 00 00
...
[517944.255600] sd 7:0:0:0: [sdd] tag#28 CDB: Write(16) 8a 00 00 00 00 00 34 03 79 00 00 00 01 b0 00 00
[517944.324232] usb 1-1.2: reset high-speed USB device number 4 using ehci-pci
[517944.481761] scsi host7: uas_eh_bus_reset_handler success
[518447.207389] sd 7:0:0:0: [sdd] tag#0 uas_eh_abort_handler 0 uas-tag 1 inflight: CMD
```

hmm, maybe related to the high IO? The io wait was 70% at the time I checked the top output during the backup. Or the controller downgraded the drive to USB2...

I did some tests with dd: 

```
## instantly after the backup
root@enterprise:/mnt/usb/backup# dd if=/dev/zero of=test.img bs=1M count=10240
10240+0 records in
10240+0 records out
10737418240 bytes (11 GB, 10 GiB) copied, 230.502 s, 46.6 MB/s

## replugged the drive
root@enterprise:/mnt/usb/backup# dd if=/dev/zero of=test2.img bs=1M count=10240
10240+0 records in
10240+0 records out
10737418240 bytes (11 GB, 10 GiB) copied, 146.327 s, 73.4 MB/s

root@enterprise:/mnt/usb/backup# dd if=/dev/zero of=test3.img bs=1M count=10240
10240+0 records in
10240+0 records out
10737418240 bytes (11 GB, 10 GiB) copied, 120.65 s, 89.0 MB/s
```

just for fun I just tried "restic check", it gives some output:

```
root@enterprise:/opt/restic/usb# restic check
password is correct
create exclusive lock for repository
load indexes
pack 8b208c32 contained in several indexes: {2ae0d498 e53ad385}
pack c4496094 contained in several indexes: {561dc9f8 ccf59786}
pack 7d2195fe contained in several indexes: {2ae0d498 e53ad385}
pack 967497df contained in several indexes: {44232fb6 cf020365}
pack 8d543a03 contained in several indexes: {44232fb6 cf020365}
pack a7b4ba37 contained in several indexes: {e0469a87 e9d08ee6}
pack 40be9a96 contained in several indexes: {2abae4cb 9fc1fbec}
...
```

Should I trust that backup? I did a rebuild then:

```
root@enterprise:/opt/restic/usb# restic rebuild-index
password is correct
counting files in repo
[2:18:07] 100.00%  461470 / 461470 packs
finding old index files
saved new index as a57847e7
remove 967 old index files
```

and checked again:

```
root@enterprise:/opt/restic/usb# restic check
password is correct
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
no errors were found
```


will try your tar cmd... and looking forward to the new archiver code ;) thanks a lot.

**Comment by fd0** [comment]
Thanks for the feedback. The kernel log mentioning `SuperSpeed` means the device is indeed accessed over USB 3, but the kernel errors are suspicious.

So, can we close this issue for now? We can always reopen it when it becomes apparent that it's a bug in restic.

**Comment by otto802** [comment]
ok, thanks anyways.

**Comment by jinie** [comment]
******
I'm unsure about the correct terminology. I refer to "Restic pack size" as the size of files stored under $RESTIC_REPOSITORY/data/[a-z0-9]{2,2} as documented [here](https://restic.readthedocs.io/en/stable/100_references.html#pack-format). 

Similarly i refer to "Borg segment size" as the files stored under $BORG_REPOSITORY/data/[a-z0-9]{1,2} as documented [here](https://borgbackup.readthedocs.io/en/stable/internals/data-structures.html#segments)
*****

    restic 0.9.4 (v0.9.4-3-g1107eef2) compiled with go1.11.4 on linux/amd64

I was going to create a new Issue when i stumbled upon this one.

I've just given up on getting Restic to successfully perform a backup on my 8TB Seagate Archive drive (SMR).
I've tested for a couple of weeks, with different filesystems, and come to the conclusion that the Restic pack size doesn't work very well with SMR drives.

My backup set consists of ~6.5TB. It has:

* 350 GB User Homes, consisting mainly of "smaller" files (along with MailDir)
* 500 GB JPEG photos.
* 1.5 TB RAW Photos
* 4+ TB of video files and other non compressible files.

I tested with Ext4, XFS and Btrfs, and for some reason Btrfs with relatime, SSD and compress=lzo options enabled gave me the best results. I assume it's because SSD's and SMR drives both need "chunks" rewritten as opposed to blocks.

With Ext4 the backup would grind to a halt around the 800GB mark. IO on the drive was 30-50MB/s, but the actual repository size grew with 2-4 MB/s. 
With XFS things got even worse, where the backup would "stall" around 600GB.
With Btrfs and SSD options, it would get to around 1.5TB in size before "stalling".

By stalling i mean growing less than 500MB in size within the range of 12 hours.

Restic appeared to be working and the only thing stalling it was IO on the drive.
A simple "ls -l" of $RESTIC_REPOSITORY/data/00 would take 2 minutes to complete.

Restic seems to prefer pack sizes of 4-6 MB.
For comparison, Borg, which uses much larger segment sizes of 500MB on the same backup set, is able to perform a full backup in a couple of days regardless of Ext4 or Btrfs being used as the filesystem.

I firmly believe that the SMR drive is what's causing the problems, as i'm doing a somewhat smaller remote backup to an Odroid HC2 equipped with a 4TB non-SMR drive, and Restic performs fine here.

I also backup the same backup set to Jottacloud via rclone restic-serve, and it also works (surprisingly) well there, so i don't doubt Restic is able to handle the amount of data.

A possible solution would be to allow larger pack sizes, and i've read lots of open and closed issues discussing the pros and cons of this. I haven't been experimenting with it, as i prefer my backup software "unmodified".

If i run the following command on my restic repository (2.2TB, non SMR)

    find $RESTIC_REPOSITORY -ls | awk '{sum += $7; n++;} END {print sprintf("%d",sum/n);}'

I get the following value : 5.016.160
For comparison, the same command on a 6TB Borg repository yields : 494.832.961

The bigger segment size ultimately results in far fewer rewrites being done despite rewriting more data.

**Comment by fd0** [comment]
Huh, thank you very much for the experience report! I don't have any hands-on experience with SMR drives yet, but I agree with your assessment that the pack size restic uses (4-6MiB) is likely the cause for the behavior you see. I'll keep this in mind when we come to implementing any sort of option to change the target size.

Overall, we may need options for users to set the target pack size, there's no one-size-fits-all value here. For users with small upstream bandwidth who use sftp or a cloud service, small files means being able to resume interrupted backups without having to re-upload much data. On the other hand, with a storage device like your SMR drive, the larger the file is the better it is for IO.

So, thanks again for the report!

**Comment by jinie** [comment]
As for a good default value, I also doubt there is a good default value, but in any case I feel the default Restic value is too low for today’s wire speed. Even when backing up over WiFi, you get 10MB/sec on a somewhat decent 802.11g connection. 

As for broadband, most connections today Re capable of uploading 6 MB in seconds. Even a 512 Kbit connection can upload 18 MB in 2 minutes. I have no idea how the rest of the world is connected, but here 100/100 mbit fiber is the slowest fiber you can buy, and if we’re talking ADSL, the slowest speed you can purchase is 50/10 mbit, which means an 18MB file will upload in 14 seconds (not taking into account async transfers)

Another consideration is that storage is growing, and prices are dropping, meaning people backup large amounts of data “for peanuts”. While filesystems can generally handle lots of files, none of them like tens of thousands of them in the same directory. Most start showing noticeable slowdowns around the 3000 files mark. For large backup sets this is a serious limiting factor, even if the drive is not SMR.

I would set an acceptable “loss rate” based on how much people can tolerate reuploading in case of resuming a backup, with a slow link as the speed variable.

So in the case of a 10 mbit upload, and lets say the acceptable loss is 1 minute, that would set the pack size to 75 MB, minus protocol overhead, so let’s say 50-60 MB.

I agree that a configurable option would be optimal, but if that’s very hard to do, perhaps the default pack size could be adjusted without too much work ? The way I read the docs, variable pack sizes are supported in a repository, meaning an existing repository would just add newer, larger, packs right along the old 4-6 MB ones.

**Comment by mholt** [comment]
(@jinie FYI my upload speed in our town is 1 Mbps.)

**Comment by fd0** [comment]
My plan was to explore if it's possible to set the pack size dynamically, based on the storage speed of the backend: Start with ~5-10MiB, measure the storage progress over a minute or so, then take that value (maybe limited by sane lower/upper bounds).
